### PR TITLE
LibWeb: Reorder table row groups according to specification

### PR DESCRIPTION
**Pull request by axgallo**
First table-header-group and table-footer-group have to be first and last, respectively, regardless of their order in the DOM.

### CODE REVIEW

File: Userland/Libraries/LibWeb/Layout/TreeBuilder.cpp [hunk 1]
```diff
@@ -733,4 +734,50 @@ void TreeBuilder::missing_cells_fixup(Vector<JS::Handle<Box>> const& table_root_
...
+        // Per https://www.w3.org/TR/css-tables-3/#table-footer-group:
+        // Like table-row-group but, for layout purposes, the first such row group is always displayed after all other rows and row groups.
+        if (footer_group)
+            row_groups.append(footer_group);
+        Vector<JS::Handle<Node>> sequence;
+        for (auto child = parent.first_child(); child; child = child->next_sibling())
+            if (is_row_group(*child))
+                sequence.append(*row_groups.take_first());
+            else
+                sequence.append(*child);
```
Comment (kalenikaliaksandr) [inline Userland/Libraries/LibWeb/Layout/TreeBuilder.cpp:776]: as I understand `sequence` is supposed to contain all parent's children when loop is done. why `row_groups.take_first()` happens for row groups? can we always do `sequence.append(*child)` instead?

File: Userland/Libraries/LibWeb/Layout/TreeBuilder.cpp [hunk 2]
```diff
@@ -733,4 +734,50 @@ void TreeBuilder::missing_cells_fixup(Vector<JS::Handle<Box>> const& table_root_
...
+        JS::Handle<Node> footer_group;
+        for (auto child = parent.first_child(); child; child = child->next_sibling()) {
+            if (!is_row_group(*child))
+                continue;
+            if (child->display().is_table_footer_group() && !footer_group)
+                footer_group = child;
+            // Accumulate table row groups and header / footer groups which aren't the first and behave like regular row groups.
+            else if (child != header_group)
+                row_groups.append(child);
+        }
```
Comment (kalenikaliaksandr) [inline Userland/Libraries/LibWeb/Layout/TreeBuilder.cpp:766]: it looks like we can collect header, footer and the rest of table groups in a single loop.

**Review by kalenikaliaksandr — CHANGES_REQUESTED** [review]

File: Userland/Libraries/LibWeb/Layout/TreeBuilder.cpp (hunk 1 above)
Comment (axgallo) [reply Userland/Libraries/LibWeb/Layout/TreeBuilder.cpp:776]: The specification doesn't say we should reorder any other children except for rows, row groups and alike (header groups and footer groups). Going by it, we have to preserve the relative order of everything else.

File: Userland/Libraries/LibWeb/Layout/TreeBuilder.cpp (hunk 2 above)
Comment (axgallo) [reply Userland/Libraries/LibWeb/Layout/TreeBuilder.cpp:766]: We also have to end up with the row groups in the correct order. I'm not sure how you'd do that cleanly with one loop.

**Comment by axgallo** [comment]
This PR addresses a real problem and the review is not aligned with the specification.

File: Userland/Libraries/LibWeb/Layout/TreeBuilder.cpp (hunk 2 above)
Comment (kalenikaliaksandr) [reply Userland/Libraries/LibWeb/Layout/TreeBuilder.cpp:766]: hmmm, why would pushing a row group into `row_groups` as you iterating children won't give you desired order? as I understand it should preserve order of everything besides header and footer nodes so I don't see why it is not possible to do in one go.

Comment (axgallo) [reply Userland/Libraries/LibWeb/Layout/TreeBuilder.cpp:766]: Feel free to take over.

**Comment by circl-lastname** [comment]
Lol?

**Pull request closed without merging**